src/Dependencies/Collections/SegmentedArray.cs

```diff
@@ -38,6 +41,315 @@ internal static void Copy<T>(SegmentedArray<T> sourceArray, SegmentedArray<T> de
+            {
+                ThrowHelper.ThrowStartIndexArgumentOutOfRange_ArgumentOutOfRange_Index();
+            }
+
+            // 2nd half of this also catches when startIndex == MAXINT, so MAXINT - 0 + 1 == -1, which is < 0.
+            if (count < 0 || startIndex - count + 1 < 0)
+            {
+                ThrowHelper.ThrowCountArgumentOutOfRange_ArgumentOutOfRange_Count();
+            }
+
+            var endIndex = startIndex - count + 1;
+            for (var i = startIndex; i >= endIndex; i--)
+            {
+                if (EqualityComparer<T>.Default.Equals(array[i], value))
+                    return i;
+            }
+
+            return -1;
+        }
+
+        public static void Reverse<T>(SegmentedArray<T> array)
+        {
+            Reverse(array, 0, array.Length);
+        }
+
+        public static void Reverse<T>(SegmentedArray<T> array, int index, int length)
+        {
+            if (index < 0)
+                ThrowHelper.ThrowIndexArgumentOutOfRange_NeedNonNegNumException();
+            if (length < 0)
+                ThrowHelper.ThrowLengthArgumentOutOfRange_ArgumentOutOfRange_NeedNonNegNum();
+            if (array.Length - index < length)
+                ThrowHelper.ThrowArgumentException(ExceptionResource.Argument_InvalidOffLen);
+
+            if (length <= 1)
+                return;
+
+            var firstIndex = index;
+            var lastIndex = index + length - 1;
+            do
+            {
+                var temp = array[firstIndex];
+                array[firstIndex] = array[lastIndex];
+                array[lastIndex] = temp;
+                firstIndex++;
+                lastIndex--;
+            } while (firstIndex < lastIndex);
+        }
+
+        public static void Sort<T>(SegmentedArray<T> array)
+        {
+            Sort(array, 0, array.Length, comparer: null);
+        }
+
+        public static void Sort<T>(SegmentedArray<T> array, int index, int length)
+        {
+            Sort(array, index, length, comparer: null);
+        }
+
+        public static void Sort<T>(SegmentedArray<T> array, IComparer<T>? comparer)
+        {
+            Sort(array, 0, array.Length, comparer);
+        }
+
+        public static void Sort<T>(SegmentedArray<T> array, int index, int length, IComparer<T>? comparer)
+        {
+            if (index < 0)
+                ThrowHelper.ThrowIndexArgumentOutOfRange_NeedNonNegNumException();
+            if (length < 0)
+                ThrowHelper.ThrowLengthArgumentOutOfRange_ArgumentOutOfRange_NeedNonNegNum();
+            if (array.Length - index < length)
+                ThrowHelper.ThrowArgumentException(ExceptionResource.Argument_InvalidOffLen);
+
+            comparer ??= Comparer<T>.Default;
+
+            var current = index;
+            foreach (var value in array.Skip(index).Take(length).OrderBy(static value => value, comparer).ToArray())
+            {
+                array[current++] = value;
+            }
```

Comment:
@Forgind this is where we need to add the new segmented sort algorithm

Reply:
The heapsort idea I mentioned would fall under the second, I guess, though I was just thinking of that because hopefully more of the new accesses would be close in the same segment or at least close to it, though that wouldn't be true for really long SegmentedArrays.

I'm also wondering about quicksort. If we assume the arrays are random already, you could quicksort everything in place in O(nlg(n)) time. Shuffling them is a little harder, but a not-quite-random shuffle should be easily doable in two passes:
1) Replace the contents of segment x with (x.Length / array.Length) of the contents of each segment, picked randomly.
2) Shuffle each segment.

I'm assuming the tiny chance of particularly bad performance isn't relevant here. The nice part of quicksort is that, since the elements you're swapping are in particular positions, you never have to go looking for an segment; you always know which to use. The obvious downside is that it isn't a stable sorting algorithm, though neither is heapsort, and I'm not sure you could get one without using extra memory or accepting worse performance.

Reply:
@Forgind for the second option, I'm thinking to use the QuickSort implementation in [ArraySortHelper](https://github.com/dotnet/runtime/blob/master/src/libraries/System.Private.CoreLib/src/System/Collections/Generic/ArraySortHelper.cs).

Reply:
This is now updated to use ArraySortHelper